### PR TITLE
TLS Options custom validation

### PR DESCRIPTION
**Pull request by usmansaleem**
Remove ArgGroups and perform custom validation for TLS options. ArgGroups cannot work from config file/environment variables.

### CODE REVIEW

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliTlsServerOptionsValidator.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    // if tls keystore is specified, the password file must be specified.
+    if ((keyStoreFile != null && keyStorePasswordFile == null)
+        || (keyStoreFile == null && keyStorePasswordFile != null)) {
+      throw new CommandLine.ParameterException(
+          spec.commandLine(),
+          "Error: --tls-keystore-file must be specified together with --tls-keystore-password-file");
+    }
+
+    final PicoCliClientAuthConstraints picoCliClientAuthConstraints =
```
Comment (rain-on) [inline commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliTlsServerOptionsValidator.java:51]: is this really a PicoCliClientAuthConstraints - or just a ClientAuthConstraints?

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliTlsServerOptions.java [hunk 2]
```diff
@@ -72,8 +73,17 @@ public File getKeyStorePasswordFile() {
...
-    return tlsClientAuthentication.tlsAllowAnyClient
-        ? Optional.empty()
-        : Optional.of(tlsClientAuthentication.clientAuthConstraints);
+    // assuming custom validation method has been called.
+
+    return tlsAllowAnyClient ? Optional.empty() : Optional.of(clientAuthConstraints);
+  }
+
+  // used by validator to enforce ArgGroup style validation
+  Boolean isTlsAllowAnyClient() {
```
[the remark this answers is not in the document]
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliTlsServerOptions.java:82]: this was a deliberate decision ... it is only meant to use with package scope.

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliTlsServerOptionsValidator.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    final File keyStoreFile = picoCliTlsServerOptions.getKeyStoreFile();
+    final File keyStorePasswordFile = picoCliTlsServerOptions.getKeyStorePasswordFile();
+
+    // no need to further validate if keystore file/password are not specified
+    if (keyStoreFile == null && keyStorePasswordFile == null) {
+      return;
+    }
+
+    // if tls keystore is specified, the password file must be specified.
+    if ((keyStoreFile != null && keyStorePasswordFile == null)
```
[the remark this answers is not in the document]
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliTlsServerOptionsValidator.java:44]: I was looking for that .. thanks :)

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliTlsServerOptions.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliTlsServerOptions.java]: removed methods.

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliTlsServerOptions.java [hunk 4]
```diff
@@ -72,8 +73,17 @@ public File getKeyStorePasswordFile() {
...
+    return tlsAllowAnyClient;
+  }
+
+  PicoCliClientAuthConstraints getPicoCliClientAuthConstraints() {
```
[the remark this answers is not in the document]
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliTlsServerOptions.java]: removed method.

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliTlsServerOptionsValidator.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.web3signer.commandline.config;
+
+import java.io.File;
+
+import picocli.CommandLine;
+
+/**
+ * Because of config file/environment default value provider, PicoCLI ArgGroup feature cannot be
```
[the remark this answers is not in the document]
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliTlsServerOptionsValidator.java]: reworded.

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliTlsServerOptionsValidator.java (hunk 1 above)
Comment (usmansaleem) [reply commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliTlsServerOptionsValidator.java]: updated.

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliTlsServerOptions.java [hunk 6]
```diff
@@ -20,45 +21,37 @@
...
-      paramLabel = MANDATORY_FILE_FORMAT_HELP,
-      required = true)
-  private File keyStorePasswordFile;
+      paramLabel = MANDATORY_FILE_FORMAT_HELP)
+  private File keyStorePasswordFile = null;
+
+  @Option(
+      names = "--tls-allow-any-client",
+      description =
+          "If set to true, any client may connect, regardless of presented certificate. This cannot "
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliTlsServerOptions.java:47]: Can the default value be included

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliClientAuthConstraints.java [hunk 7]
```diff
@@ -32,8 +33,9 @@
 
   @Option(
       names = "--tls-allow-ca-clients",
-      description = "If defined, allows clients authorized by the CA to connect to EthSigner.",
-      arity = "0")
+      description = "If set to true, allows clients authorized by the CA to connect to Web3Signer.",
+      paramLabel = MANDATORY_BOOL_FORMAT_HELP,
+      arity = "1")
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliClientAuthConstraints.java:38]: Can the default value be included

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/Web3SignerBaseCommand.java [hunk 8]
```diff
@@ -223,7 +222,11 @@ public String getMetricsNetworkInterface() {
 
   @Override
   public Optional<TlsOptions> getTlsOptions() {
-    return Optional.ofNullable(picoCliTlsServerOptions);
+    if (picoCliTlsServerOptions.getKeyStoreFile() != null
+        && picoCliTlsServerOptions.getKeyStorePasswordFile() != null) {
+      return Optional.of(picoCliTlsServerOptions);
+    }
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/web3signer/commandline/Web3SignerBaseCommand.java:228]: Might be nice if the picoCliTlsServerOptions returned a Optional<TlsOptions> so this logic could be moved out the base command and stay with the TLS options

File: commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliTlsServerOptionsValidator.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    final File keyStoreFile = picoCliTlsServerOptions.getKeyStoreFile();
+    final File keyStorePasswordFile = picoCliTlsServerOptions.getKeyStorePasswordFile();
+
+    // no need to further validate if keystore file/password are not specified
+    if (keyStoreFile == null && keyStorePasswordFile == null) {
+      return;
+    }
+
+    // if tls keystore is specified, the password file must be specified.
+    if (onlyOneInitialized(keyStoreFile, keyStorePasswordFile)) {
```
Comment (jframe) [inline commandline/src/main/java/tech/pegasys/web3signer/commandline/config/PicoCliTlsServerOptionsValidator.java:46]: Can you add some tests if there aren't already some existing tests that cover these validation checks

**Review by rain-on — APPROVED** [review]
LGTM, but would be nice to remove unused variable.

**Pull request merged**